### PR TITLE
Documentation

### PR DESCRIPTION
**Pull request by jounathaen**
I tried to improve the documentation a bit. (#22)

Please review, as I'm new to this crate and I might have misunderstood things.
The example is adapted from the serial example.

### CODE REVIEW

File: src/serial.rs [hunk 1]
```diff
@@ -101,7 +140,13 @@ macro_rules! hal {
         ),
     )+) => {
         $(
+            $(#[$meta])*
             impl<PINS> Serial<$USARTX, PINS> {
+
+            /// Configures the Serial interface and creates an interface struct.
+            /// `baud_rate` configures the baud rate of the interface. The other parameters are
+            /// handles to the corresponding registers, to make Rust's borrow checker happy.
```
Comment (TheZoq2) [inline src/serial.rs]: The regs aren't passed to make the borrow checker happy, they are passed to be modified in order to make the device work.

Comment (TheZoq2) [inline src/serial.rs]: The regs aren't passed to make the borrow checker happy, they are passed to be modified in order to make the device work.

File: src/serial.rs [hunk 2]
```diff
@@ -101,7 +140,13 @@ macro_rules! hal {
...
+            /// (Ownership must be passed to this function)
```
Comment (TheZoq2) [inline src/serial.rs]: I would prefer "Takes ownership of the usart device" or something along those lines

File: src/serial.rs [hunk 3]
```diff
@@ -141,24 +186,31 @@ macro_rules! hal {
                     Serial { usart, pins }
                 }
 
+                /// Starts listening by enabling _RX buffer not empty_ interrupt and _TX register
+                /// empty_ interrupt
```
Comment (TheZoq2) [inline src/serial.rs]: "RX buffer not empty" does not sound right. The interrupts are just "tx and rx interrupts"

File: src/serial.rs [hunk 4]
```diff
@@ -403,6 +458,7 @@ hal! {
         |remap| remap == 1,
         APB2
     ),
+    #[doc(hidden)]
```
Comment (TheZoq2) [inline src/serial.rs]: I don't like hiding the docs for the other devices. A comment similar to that for USART1 would be better imo

File: src/serial.rs [hunk 5]
```diff
@@ -1,3 +1,41 @@
+//! # Serial Communication
+//! This module contains the functions to utilize the USART (Universal
+//! synchronous asynchronous receiver transmitter (USART)
```
Comment (eldruin) [inline src/serial.rs]:
```suggestion
//! synchronous asynchronous receiver transmitter)
```
Small duplication here.

File: src/serial.rs [hunk 6]
```diff
@@ -1,3 +1,41 @@
...
+//!
+//!
+//! ## Example usage:
+//! ```rust
+//!// prelude: create handles to the peripherals and registers
```
Comment (eldruin) [inline src/serial.rs]: I would add at least one space between //! and the code for reading ease when reading the documentation un-rendered.

File: src/serial.rs (hunk 1 above)
Comment (jounathaen) [reply src/serial.rs]: Yes, but without the borrow checker, the registers could be encoded in the function statically (C-style), or have I misunderstood the concept of the peripheral struct?

I thought, that this is the reason that 6 parameters are passed to that function, whilst only one is a real parameter.

Comment (jounathaen) [reply src/serial.rs]: (But yes, that is sloppy language)

File: src/serial.rs (hunk 3 above)
Comment (jounathaen) [reply src/serial.rs]: I now use the names from the datasheet:
- Transmit data register empty
- Received data ready to be read

File: src/serial.rs [hunk 7]
```diff
@@ -101,7 +140,15 @@ macro_rules! hal {
...
         $(
+            $(#[$meta])*
+            /// The behaviour of the functions is equal for all three USARTs.
+            /// Except that they are using the corresponding USART hardware and pins.
             impl<PINS> Serial<$USARTX, PINS> {
+
+                /// Configures the Serial interface and creates an interface struct.
+                /// `baud_rate` configures the baud rate of the interface. The other parameters are
+                /// handles to the corresponding registers which must be changed by this function.
+                /// This function takes ownership of these registers.
```
Comment (TheZoq2) [inline src/serial.rs]: This isn't 100% correct for two reasons:

- Not all other parameters are registers, `Clocks` is a struct which 1. indicates that the clock configuration is frozen and 2, gives the frequencies of the various clocks. Also, pins are not registers, they are input and output pins
- It only takes ownership of the pins and usart device registers, the other registers are passed by reference in order to initialise the device

**Comment by TheZoq2** [comment]
Thanks for the updates! 

Apart from the serial init comment, it looks good to me now

File: src/serial.rs (hunk 7 above)
Comment (jounathaen) [reply src/serial.rs]: Thanks for the clarification. I adapted it accordingly to your explanation and I learned a lot about the functionality :+1:

Comment (TheZoq2) [reply src/serial.rs]: Looks almost perfect now!

> The `serial` struct takes ownership of the `PINS` and `USARTX` device registers.

I read this as "takes ownership of the `PINS` device registers and `USARTX` device registers which is incorrect, `PINS` is a tuple of individual pins. Perhaps changing it to "Takes ownership over the `USARTX` device registers and the specified `PINS`" would make it more clear.

> The `serial` struct does not take ownership of `MAPR` and `APBX`

Another, more style related nitpick is that I'm not sure if this comment is required as that can be infered from them being passed as references. 

Other than those two things, it looks good to me

Comment (jounathaen) [reply src/serial.rs]: As you wish, master!

**Comment by TheZoq2** [comment]
Looks perfect now, great work! And thanks for being patient with my nitpicks :)

@therealprof I'd say this is ready to merge. Just to be sure since I've never done this before, do I do a merge or a squash? It looks like #20 was squashed but I can't tell for sure

**Comment by therealprof** [comment]
@TheZoq2 Squashing sounds like a good idea. I'd only keep the individual commits if they make sense on their own but these are mostly fixes on top of other commits.

**Pull request merged**